### PR TITLE
Serialize prettytoml ArrayElements as lists

### LINKED ISSUE #2504: Can't lock package with `git` and `extras`
##### Issue description

A package spec with `git` and `extras` fails to lock. This happens regardless of the `editable` flag.

##### Expected result

The package should be installed with all extras.

##### Actual result

```
[I] ~/test $ pipenv --version
pipenv, version 2018.7.1
[I] ~/test $ ll
total 8
-rw-r--r--  1 jacobhayes    96B Jul  3 13:12 Pipfile
[I] ~/test $ cat Pipfile
[packages]
requests = {git = "https://github.com/requests/requests.git", extras = ["security"]}
[I] ~/test $ pipenv --rm ; pipenv install --verbose
Removing virtualenv (/Users/jacobhayes/.local/share/virtualenvs/test-qFl43Lkt)...
Creating a virtualenv for this project...
Pipfile: /Users/jacobhayes/test/Pipfile
Using /usr/local/opt/python/bin/python3.6 (3.6.5) to create virtualenv...
⠋Already using interpreter /usr/local/opt/python/bin/python3.6
Using base prefix '/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6'
New python executable in /Users/jacobhayes/.local/share/virtualenvs/test-qFl43Lkt/bin/python3.6
Also creating executable in /Users/jacobhayes/.local/share/virtualenvs/test-qFl43Lkt/bin/python
Installing setuptools, pip, wheel...done.
Setting project for test-qFl43Lkt to /Users/jacobhayes/test

Virtualenv location: /Users/jacobhayes/.local/share/virtualenvs/test-qFl43Lkt
Pipfile.lock not found, creating...
Locking [dev-packages] dependencies...
Locking [packages] dependencies...
Traceback (most recent call last):
  File "/usr/local/bin/pipenv", line 11, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/pipenv/cli.py", line 435, in install
    selective_upgrade=selective_upgrade,
  File "/usr/local/lib/python3.6/site-packages/pipenv/core.py", line 1943, in do_install
    pypi_mirror=pypi_mirror,
  File "/usr/local/lib/python3.6/site-packages/pipenv/core.py", line 1312, in do_init
    pypi_mirror=pypi_mirror,
  File "/usr/local/lib/python3.6/site-packages/pipenv/core.py", line 1140, in do_lock
    project.write_lockfile(lockfile)
  File "/usr/local/lib/python3.6/site-packages/pipenv/project.py", line 648, in write_lockfile
    content, indent=4, separators=(u',', u': '), sort_keys=True,
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/encoder.py", line 430, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/encoder.py", line 404, in _iterencode_dict
    yield from chunks
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/encoder.py", line 404, in _iterencode_dict
    yield from chunks
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/encoder.py", line 404, in _iterencode_dict
    yield from chunks
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/encoder.py", line 437, in _iterencode
    o = _default(o)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/encoder.py", line 180, in default
    o.__class__.__name__)
TypeError: Object of type 'ArrayElement' is not JSON serializable
```

##### Steps to replicate

```toml
[packages]
requests = {git = "https://github.com/requests/requests.git", extras = ["security"]}
```

`pipenv install`

-------------

<details><summary>$ pipenv --support</summary>

Pipenv version: `'2018.7.1'`

Pipenv location: `'/usr/local/lib/python3.6/site-packages/pipenv'`

Python location: `'/usr/local/opt/python/bin/python3.6'`

Other Python installations in `PATH`:

  - `2.7`: `/usr/local/bin/python2.7`
  - `2.7`: `/usr/local/bin/python2.7`
  - `2.7`: `/usr/bin/python2.7`
  - `3.6`: `/usr/local/bin/python3.6m`
  - `3.6`: `/usr/local/bin/python3.6`

  - `2.7.15`: `/usr/local/bin/python`
  - `2.7.10`: `/usr/bin/python`
  - `2.7.15`: `/usr/local/bin/python2`
  - `3.6.5`: `/usr/local/bin/python3`

PEP 508 Information:

```
{'implementation_name': 'cpython',
 'implementation_version': '3.6.5',
 'os_name': 'posix',
 'platform_machine': 'x86_64',
 'platform_python_implementation': 'CPython',
 'platform_release': '17.6.0',
 'platform_system': 'Darwin',
 'platform_version': 'Darwin Kernel Version 17.6.0: Tue May  8 15:22:16 PDT '
                     '2018; root:xnu-4570.61.1~1/RELEASE_X86_64',
 'python_full_version': '3.6.5',
 'python_version': '3.6',
 'sys_platform': 'darwin'}
```

System environment variables:

  - `Apple_PubSub_Socket_Render`
  - `COLORFGBG`
  - `COLORTERM`
  - `COPYFILE_DISABLE`
  - `EDITOR`
  - `GOPATH`
  - `HOME`
  - `ITERM_PROFILE`
  - `ITERM_SESSION_ID`
  - `LANG`
  - `LOGNAME`
  - `PAGER`
  - `PATH`
  - `PWD`
  - `SECURITYSESSIONID`
  - `SHELL`
  - `SHLVL`
  - `SSH_AUTH_SOCK`
  - `TERM`
  - `TERM_PROGRAM`
  - `TERM_PROGRAM_VERSION`
  - `TERM_SESSION_ID`
  - `TMPDIR`
  - `TMUX`
  - `TMUX_PANE`
  - `TMUX_PLUGIN_MANAGER_PATH`
  - `USER`
  - `XPC_FLAGS`
  - `XPC_SERVICE_NAME`
  - `__CF_USER_TEXT_ENCODING`
  - `PYTHONDONTWRITEBYTECODE`
  - `PIP_PYTHON_PATH`

Pipenv–specific environment variables:

Debug–specific environment variables:

  - `PATH`: `/Users/jacobhayes/bin:/Users/jacobhayes/bin:/usr/local/Caskroom/google-cloud-sdk/latest/google-cloud-sdk/bin:/Users/jacobhayes/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin`
  - `SHELL`: `/usr/local/bin/fish`
  - `EDITOR`: `vim`
  - `LANG`: `en_US.UTF-8`
  - `PWD`: `/Users/jacobhayes/test`

---------------------------

Contents of `Pipfile` ('/Users/jacobhayes/test/Pipfile'):

```toml
[packages]
requests = {git = "https://github.com/requests/requests.git", extras = ["security"]}

```

</details>

### PR DESCRIPTION
**Pull request by JacobHayes**
Applies the fix suggested in https://github.com/pypa/pipenv/issues/2504#issuecomment-402853014

Resolves #2504

I couldn't find a quick/easy way to test this locally before pushing (just saw the Makefile), so using buildkite to test a bit blind. 😅

### CODE REVIEW

File: tests/integration/test_lock.py [hunk 1]
```diff
@@ -348,6 +347,28 @@ def test_lock_editable_vcs_without_install(PipenvInstance, pypi):
...
+            """.strip())
+        c = p.pipenv('lock')
+        assert c.return_code == 0
+        assert 'requests' in p.lockfile['default']
+        assert 'idna' in p.lockfile['default']
+        assert 'chardet' in p.lockfile['default']
+        assert 'cryptography' in p.lockfile['default']
+        assert 'pyOpenSSL' in p.lockfile['default']
+        c = p.pipenv('install')
+        assert c.return_code == 0
```
Comment (JacobHayes) [inline tests/integration/test_lock.py]: This is largely a copy of `test_lock_editable_vcs_without_install` above, adding the `extras` mark, swapping the `ref` for `extras` in the spec, and adding checks for the security packages.

**Comment by JacobHayes** [comment]
I can't see the output from buildkite and running tests locally with `make test` fails for other reasons (will try to work through and add the fixes). I'm guessing something in my added test was broken because the buildkite tests passed before I added this test (and reordered/rebased).

If anyone that can see the error could post it back, much appreciated.

**Comment by techalchemy** [comment]
As an FYI `make test` doesn't do anything as far as I'm aware? If you want to test locally you just need to run 

```console
pip install -e .
pipenv install --dev
pipenv run pytest -v -n auto --ignore pipenv/patched --ignore pipenv/vendor tests
```

To run a specific test you can replace the last command with `pipenv run pytest -v -n auto -k 'test_function_name'`

Buildkite output:

```
[gw4] linux2 -- Python 2.7.15 /root/.local/share/virtualenvs/pipenv-wInVs968-2.7/bin/python2.7
 
PipenvInstance = <class 'tests.integration.conftest._PipenvInstance'>
pypi = <pytest_pypi.serve.Server object at 0x7f0db1ffc590>
 
    @pytest.mark.extras
    @pytest.mark.lock
    @pytest.mark.vcs
    @pytest.mark.needs_internet
    def test_lock_editable_vcs_with_extras_without_install(PipenvInstance, pypi):
        with PipenvInstance(pypi=pypi, chdir=True) as p:
            with open(p.pipfile_path, 'w') as f:
                f.write("""
    [packages]
    requests = {git = "https://github.com/requests/requests.git", editable = true, extras = ["security"]}
                """.strip())
            c = p.pipenv('lock')
>           assert c.return_code == 0
E           AssertionError: assert 1 == 0
E            +  where 1 = <Command 'pipenv lock'>.return_code
 
tests/integration/test_lock.py:362: AssertionError

----------------------------- Captured stdout call -----------------------------
$ pipenv lock
 
Creating a virtualenv for this project…
Pipfile: /tmp/pipenv-5pAYbh-project/Pipfile
Using /root/.local/share/virtualenvs/pipenv-wInVs968-2.7/bin/python2.7 (2.7.15) to create virtualenv…
Already using interpreter /root/.local/share/virtualenvs/pipenv-wInVs968-2.7/bin/python2.7
Using real prefix '/usr'
New python executable in /tmp/pipenv-5pAYbh-project/.venv/bin/python2.7
Also creating executable in /tmp/pipenv-5pAYbh-project/.venv/bin/python
Installing setuptools, pip, wheel...done.
 
Virtualenv location: /tmp/pipenv-5pAYbh-project/.venv
Locking [dev-packages] dependencies…
Locking [packages] dependencies…
 
CRITICAL:notpip._internal.index:Could not find a version that satisfies the requirement pyOpenSSL>=0.14 (from requests[security]->-r /tmp/pipenv-ebQyJY-requirements/pipenv-xdUlRl-constraints.txt (line 2)) (from versions: )
Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
  You can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.
  Hint: try $ pipenv lock --pre if it is a pre-release dependency.
No matching distribution found for pyOpenSSL>=0.14 (from requests[security]->-r /tmp/pipenv-ebQyJY-requirements/pipenv-xdUlRl-constraints.txt (line 2))
```

**Comment by JacobHayes** [comment]
@techalchemy Thanks for the testing help and failure output. The test should be fixed now. I've run it locally to verify the test case triggers the intended error (since the code fix was committed before the test fix) by rebasing out the code fix commit, testing (and verifying if fails with `TypeError: Object of type 'ArrayElement' is not JSON serializable`, then force pulling, and testing again.

I documented a lot of my difficulties because I'd love for it to be easier to contribute to this awesome project. I'll post a follow up PR mentioned below to fix some of the more procedural issues, as opposed to more structural decisions like pypa/pipenv#2557 (which you hint at with `pip install -e .` before testing) and the Makefile/Docker changes.

`make test` will run `run-tests.sh` (via docker-compose), but the linked image doesn't have python 3.7, so fails those tests. The Dockerfile is outside this repo, so can't be updated as readily. `docker-compose up` is usually used to start a service (that intends to keep running) as opposed to a command, which is usually run with `docker-compose run {service} {command}`, so the usage in the `Makefile` is a bit obscured.

I had tried to run `pytest`, but that brought in ~300 tests because the `build` directory after pip installing had them, which wasn't included in `pytest.ini` after `norecursedir` was rightly overridden to include the `vendor` and `patched` exclusions you listed. After updating `pytest.ini` to add back the defaults, an argument-less `pytest` had the same number of tests as your posted command above (185). The tests still take awhile to run, so running an individual test is nice, but see below on a caveat.

It's unclear how to add a test for a Pipfile with a new dependency. It's not particularly obvious that a local pypi clone is being used and new packages need to be added to `tests/pypi/`. Even after adding them, tests have to be run from the root repo directory for the local registry to find the packages (ex: can't run pytest from `tests/integration`) due to a bug in the pytest-pypi package or how it's passed `PYPI_VENDOR_DIR`. (Since tests take so long, I often try to run them from the folder containing the test I've just edited.)

I'll submit a follow up PR to fix the `pytest-pypi` package and `pytest.ini` and can add some testing documentation in `CONTRIBUTING.md` about how to add packages, etc. For the `Makefile`, it'd probably be nice to move the testing Dockerfile to this repo as `Dockerfile.test`, add python 3.7, and change the `docker-compose.yml` to build from that. Then maybe rename the `test` target to `test-all` and use `docker-compose run` instead of `up` and add a simpler `test` target that just runs `pytest`.

I want to point out how appreciative I am of this project and it's maintainers; it is now so much easier to manage dependencies, and I'd love to see this project continue to grow by lowering the barrier to contributing.

**Comment by techalchemy** [comment]
there are a lot of words here that I'm not really sure are relevant to the PR. Is this good to go in your view?

**Comment by JacobHayes** [comment]
Haha yes, it's good on my end. I'll move the relevant details in that comment to pypa/pipenv#2568.

**Comment by techalchemy** [comment]
Awesome thanks. I put my specific comments there but am super appreciative of your effort and am excited to get those updates merged!

**Comment by techalchemy** [comment]
Ah there is a failure right now also:

```console
assert 'pysocks' in p.lockfile['default']
AssertionError: assert 'pysocks' in {'certifi': {'hashes': ['sha256:13e698f54293db9f89122b0581843a782ad0934a4fe0172d2a980ba77fc61bb7', 'sha256:9fa520c1bac... ['socks'], 'git': 'https://github.com/requests/requests.git', 'ref': '4499401a55dedea27831fd9b3c6249b9679f3a2d'}, ...}
```

seems this is now called `socks`?

**Comment by JacobHayes** [comment]
Seems the sub-dep is no longer frozen and is instead stored in the package `extras`. I forget whether Pipenv is supposed store hashes of sub deps of editable paths, but regardless the fix should be pushed.

**Comment by uranusjr** [comment]
The idea looks correct to me, but I wonder if we should add more types to check, or maybe check against one of the `collections.abc` types instead.

**Comment by frostming** [comment]
The fix is OK, though a bit dirty, may be better if `ArrayElement` inherits `list` so we don't need to hack `json.dumps`

**Comment by JacobHayes** [comment]
Having `ArrayElement` subclass `list` would work alright, but similar would also need to be done for `AbstractTable` (`dict`-likes) and `AtomicElement` (`int`, `str`, `date`, etc). The latter wouldn't quite work since it represents multiple distinct python types.

I don't think checking against a `collections.abc` helps since `primitive_value` is not a standard attribute (unless we were to try to recreate the primitive_value based on the ABC, which is what `primitive_value` does for us, but again wouldn't work for `AtomicElement`). However, that leads us to just checking for `hasattr(obj, 'primitive_value')`, which should work for any of the mentioned prettytoml types listed above.

**Comment by uranusjr** [comment]
@JacobHayes I tried to make the encoding process cleaner with a custom encoder class, and check against a real base class instead of relying on duck-typing. Could you confirm if this solution works for you?

**Comment by JacobHayes** [comment]
@uranusjr Yup, that works for me. As is, it wouldn't cover TokenElements (ex: AtomicElement for str/int/basic types), but it doesn't seem like any of those filter through afaik. I think this sufficently solves the issue, but perhaps another question worth asking is why do ArrayElements filter through the codebase when basic ones do not (ie: is there a literal leaky abstraction somewhere)?

**Comment by techalchemy** [comment]
thanks for the cleanup on this and for allowing me to continue to be ignorant in this area of the codebase. This is particularly appreciated!

**Pull request merged**